delivery_auto_refresh/models/sale_order.py

```diff
@@ -56,63 +85,109 @@ def _get_param_auto_add_delivery_line(self):
+                # exists, substract 1
+                if delivery_line[f] != val - 1:
+                    new_vals[f] = val
+            elif delivery_line[f] != val:
+                new_vals[f] = val
+        if new_vals:
+            delivery_line.write(new_vals)
+
     def _auto_refresh_delivery(self):
         self.ensure_one()
-        # Make sure that if you have removed the carrier, the line is gone
-        if self.state in {"draft", "sent"}:
-            # Context added to avoid the recursive calls and save the new
-            # value of carrier_id
-            self.with_context(auto_refresh_delivery=True)._remove_delivery_line()
-        if self._get_param_auto_add_delivery_line() and self.carrier_id:
-            if self.state in {"draft", "sent"}:
-                price_unit = self.carrier_id.rate_shipment(self)["price"]
-                if not self.is_all_service:
-                    self._create_delivery_line(self.carrier_id, price_unit)
-                self.with_context(auto_refresh_delivery=True).write(
-                    {"recompute_delivery_price": False}
-                )
-
-    @api.model
-    def create(self, vals):
-        """Create or refresh delivery line on create."""
-        order = super().create(vals)
-        order._auto_refresh_delivery()
-        return order
+        if self.env.context.get("auto_refresh_delivery"):
+            return
+        if not self._get_param_auto_add_delivery_line():
+            return
+        if self.state not in ("draft", "sent"):
+            return
+
+        self = self.with_context(auto_refresh_delivery=True)
+
+        if not self.carrier_id:
+            self._set_delivery_carrier()
+
+        if not self.carrier_id:
+            self._remove_delivery_line()
+        elif self.is_all_service:
+            self._remove_delivery_line()
+        else:
+            price_unit = self.carrier_id.rate_shipment(self)["price"]
+            delivery_lines = self.order_line.filtered("is_delivery")
+            if not delivery_lines:
+                self._create_delivery_line(self.carrier_id, price_unit)
+            elif len(delivery_lines) > 1:
+                delivery_discount = self._get_delivery_discount()
+                self._remove_delivery_line()
+                sol = self._create_delivery_line(self.carrier_id, price_unit)
+                if delivery_discount and sol:
+                    sol.discount = delivery_discount
+            else:
+                self._update_delivery_line(delivery_lines[0], price_unit)
+        if self.recompute_delivery_price:
+            self.recompute_delivery_price = False
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # Prevent to refresh delivery in the call to super
+        orders = (
+            super(SaleOrder, self.with_context(auto_refresh_delivery=True))
+            .create(vals_list)
+            .with_context(auto_refresh_delivery=False)
+        )
+        for order in orders:
```

Comment:
```suggestion
        )
        for order in orders.with_context(auto_refresh_delivery=False):
```
I'd find this more clear. Matter of taste likely....

Reply:
I prefer my way as it is linked to the call to create where I set context before and remove after...